docs/code-reviewing.md

```diff
@@ -0,0 +1,99 @@
+# Reviewing Zulip server code
+
+This document is a brief discussion of what we look for when reviewing
+contributions to Zulip. It's meant partially for developers who want
+to get their code merged faster, and partially for developers who have
+made successful pull requests already and would like to start
+participating in code review.
+
+* *The Travis CI build.* The tests need to pass. One can investigate
+  any failures and figure out what to fix by clicking on a red X next
+  to the commit hash or the Detail links on a pull request. (Example:
+  in [#1219](https://github.com/zulip/zulip/pull/1219), click the red
+  X next to `f1f474e` to see the build jobs for that commit, at least
+  one of which has failed. Click on the link for Travis continuous
+  integrations details to see [the tests Travis ran on that
+  commit](https://travis-ci.org/zulip/zulip/builds/144300899), at
+  least one of which failed, and go to [one of the failing
+  tests](https://travis-ci.org/zulip/zulip/jobs/144300901) to see the
+  error.)
+
+* *Technical design.* There are a lot of considerations here:
+  security, migration paths/backwards compatibility, cost of new
+  dependencies, interactions with features, speed of performance, API
+  changes.
+
+* *User interface and visual design.* If frontend changes are
+  involved, the reviewer will check out the code, play with the new
+  UI, and verify it for both quality and consistency with the rest of
+  the Zulip UI.  We highly encourage posting screenshots to save
+  reviewers time in getting a feel for what the feature looks like --
+  you'll get a quicker response that way.
+
+* *Error handling.* The code should always check for invalid user
+  input.  User-facing error messages should be clear and when possible
+  be actionable (it should be obvious to the user what they need to do
+  in order to correct the problem).
+
+* *Testing.* The tests should validate that the feature works
+  correctly, and specifically test for common error conditions, bad
+  user input, and potential bugs that are likely for the type of
+  change being made.  Tests that exclude whole classes of potential
+  bugs are preferred when possible (e.g. the common test suite between
+  the frontend and backend markdown processors or the `GetEventsTest`
+  test for buggy race condition handling).  If the feature involves
+  frontend changes, there should be frontend tests.  See the [test
+  writing][test-writing] documentation for more details.
```

Comment:
Add to this we are trying to maintain ~100% test coverage on the backend, so backend changes should have negative tests for the various error conditions.

Also we should link to the relevant part of markdown.md for the "common test suite" thing.